File: lib/python/a_fun_xam.py

# This file can be automatically generaeted using the following command
# m4 ../python.m4 ../xam/a_fun_xam.m4 > a_fun_xam.py
# -----------------------------------------------------------------------------
#         cppad_swig: A C++ Object Library and Swig Interface to CppAD
#          Copyright (C) 2017-17 Bradley M. Bell (bradbell@seanet.com)
#              This program is distributed under the terms of the
#          GNU Affero General Public License version 3.0 or later see
#                     http://www.gnu.org/licenses/agpl.txt
# -----------------------------------------------------------------------------
# std::vector<a_double>
# -----------------------------------------------------------------------------
def a_fun_xam() :
	#
	# load the cppad Swig library
	import py_cppad
	#
	# initialize return variable
	ok = True
	n = 2
	m = 1
	#
	# create ax
	x = py_cppad.vector_double(n)
	for i in range( n  ) :
		x[i] = i + 1.0
	ax = py_cppad.independent(x)
	#
	# create af
	ax0 = ax[0]
	ax1 = ax[1]
	ay = py_cppad.vector_ad(1)
	ay[0] = ax0 + ax0 - ax1
	af = py_cppad.a_fun(ax, ay)
	#
	# zero order forward
	x[0] = 3.0
	x[1] = 1.0
	y = af.forward(0, x)
	ok = ok and y[0] == 5.0
	#
	# first order forward
	x[0] = 0.0
	x[1] = 1.0
	y = af.forward(1, x)
	ok = ok and y[0] == -1.0
	#
	return( ok )
#

File: lib/cplusplus/a_fun_xam.cpp
// This file can be automatically generaeted using the following command
// m4 ../cpp.m4 ../xam/a_fun_xam.m4 > a_fun_xam.cpp
// -----------------------------------------------------------------------------
//         cppad_swig: A C++ Object Library and Swig Interface to Cppad
//          Copyright (C) 2017-17 Bradley M. Bell (bradbell@seanet.com)
//              This program is distributed under the terms of the
//          GNU Affero General Public License version 3.0 or later see
//                     http://www.gnu.org/licenses/agpl.txt
// -----------------------------------------------------------------------------
// std::vector<a_double>
// -----------------------------------------------------------------------------
# include <cstdio>
# include <cppad/swig/a_double.hpp>
# include <cppad/swig/a_fun.hpp>
# include <cppad/swig/function.hpp>
# include <cppad/swig/typedef.hpp>

bool a_fun_xam(void) {
	using cppad_swig::a_double;
	//
	// initialize return variable
	bool ok = true;
	size_t n = 2;
	//
	// create ax
	vector_double x = vector_double(n);
	for(size_t i = 0; i < n ; i++) {
		x[i] = i + 1.0;
	}
	vector_ad ax = cppad_swig::independent(x);
	//
	// create af
	a_double ax0 = ax[0];
	a_double ax1 = ax[1];
	vector_ad ay = vector_ad(1);
	ay[0] = ax0 + ax0 - ax1;
	cppad_swig::a_fun af = cppad_swig::a_fun(ax, ay);
	//
	// zero order forward
	x[0] = 3.0;
	x[1] = 1.0;
	vector_double y = af.forward(0, x);
	ok = ok && y[0] == 5.0;
	//
	// first order forward
	x[0] = 0.0;
	x[1] = 1.0;
	y = af.forward(1, x);
	ok = ok && y[0] == -1.0;
	//
	return( ok );
}
